cvat-data/src/js/cvat-data.js

```diff
@@ -116,56 +119,96 @@ class FrameProvider {
         @param start {number} - is the first frame of a block
         @param end {number} - is the last frame of a block + 1
+        @param needed_frame {number} - number of frame that should be returned as soon as
+        @param callback - callback)
+
     */
-    decode(block, start, end) {
-        async function decode(resolve, reject) {
-            // no more requests possible during one is running
+
+    start_decode(block, start, end, needed_frame, callback)
+    {
+         if (this._blockType === BlockType.TSVIDEO){
             if (this._running) {
                 throw new Error('Decoding has already running');
             }
-
-            try {
-                this._running = true;
-                this._blocks.push(`${start}:${end}`);
-
-                this._cleanup();
-
-                if (this._blockType === BlockType.TSVIDEO) {
-                    this._demuxer.write(block);
-                    for (let i = start; i < end; i++) {
-                        const result = this._videoDecoder.decode();
-                        if (!Array.isArray(result)) {
-                            const message = 'Result must be an array.'
-                                + `Got ${result}. Possible reasons: `
-                                + 'bad video file, unpached jsmpeg';
-                            throw Error(message);
-                        }
-                        this._frames[i] = YCbCrToRGBA(...result);
-                    }
-                    resolve();
-                } else if (this._blockType === BlockType.ARCHIVE) {
-                    // to do with any npm dearchiver
-                }
-            } catch (error) {
-                reject(error);
-            } finally {
-                this._running = false;
+            this._demuxer.write(block);
+            this._blocks.push(`${start}:${end}`);
+            this._cleanup();
+            this._currFrame = start;
+            for (var i = start; i < end; i++)
+            {
+                this._frames[i] = "loading";
             }
-        }
+            this._running = true;
+            this._decode_id = setInterval(this.decode.bind(this), this, start, end, needed_frame, callback, 10);
+           
+        }else{
+            var zip = new JSZip();
+            var that = this;
+            zip.loadAsync(block).then(function(_zip){
+                var index = start;
+               
+                _zip.forEach(function (relativePath, zipEntry) { 
+                    
+                   _zip.file(relativePath).async('ArrayBuffer').then(function (fileData){     
+                          // let u8 = new Uint8Array(fileData);
+                          // let b64encoded = btoa([].reduce.call(new Uint8Array(fileData),function(p,c){return p+String.fromCharCode(c)},''));
+                          // let mimetype="image/jpeg";
+                          that._frames[index] = jpeg.decode(fileData); //"data:"+mimetype+";base64,"+b64encoded;
```

Comment:
What if input type isn't jpeg?

Reply:
And how long does decoding work?

Reply:
> What if input type isn't jpeg?

Yes, this is problem. I have only array of bytes, not base64. And i do not known how to decode this...